design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,140 @@
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost API v1beta1
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes making certain backward incompatible changes to the
+BareMetalHost API and publishing the result as version `v1beta1`.
+
+## Motivation
+
+The BareMetalHost API is currently stuck at `v1alpha1`. Meanwhile, several
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- Clean up issues in the BareMetalHost API that have accumulated during
+  the usage of `v1alpha1`.
+
+- Make the API version reflect the actual support status of the API (see
+  [alternatives](#Alternatives) for details).
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API completely (as `v1`).
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, BMO will use its own architecture as the default.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Add `Architecture`
+- Refactor the BMO code to avoid referring directly to the current version.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Communicate the change to the community.
+
+### Dependencies
+
+None
+
+### Test Plan
+
+We'll need to test both API versions. Since we already have tests for different
+versions of CAPM3, we may tie them to different versions of BMH.
```

Comment:
I would like to know a bit more concrete idea about how this would be incorporated in the test. Are we thinking about creating new branch. Should this be tied to the issue https://github.com/metal3-io/baremetal-operator/issues/1283 ? How CI would accomodate this greatly depends on how we would want to land this new API. For CAPM3 we have different branches currently for each minor release.